### ISSUE TITLE
schema-codegen for typescript, build failed by TS1255 - and other minor linting errors

### ISSUE THREAD

**Issue by AlexFBP** [colyseus/schema#103]
First of all, thanks for the whole Colyseus Project

`schema-codegen path/to/room.ts --ts` generated, among others, the following file:

```typescript
// 
// THIS FILE HAS BEEN GENERATED AUTOMATICALLY
// DO NOT CHANGE IT MANUALLY UNLESS YOU KNOW WHAT YOU'RE DOING
// 
// GENERATED USING @colyseus/schema 1.0.6
// 

import { Schema, type, ArraySchema, MapSchema, DataChange } from "@colyseus/schema";
import { Player } from "./Player"

export class GameState extends Schema {
    @type({ map: Player }) public players!: MapSchema<Player> = new MapSchema<Player>();
    @type("boolean") public running!: boolean;
}
```

but when using this within an Angular (10) development, as it comes, in terminal appears the following:
```log
    ERROR in src/app/genlib/GameState.ts:12:42 - error TS1255: A definite assignment assertion '!' is not permitted in this context.
    
    12     @type({ map: Player }) public players!: MapSchema<Player> = new MapSchema<Player>();
                                                ~
    
```

I "fixed" by manually editing to remove the `!` like this...:
```diff
@@ -9,6 +9,6 @@ import { Schema, type, ArraySchema, MapSchema, DataChange } from "@colyseus/sche
 import { Player } from "./Player"
 
 export class GameState extends Schema {
-    @type({ map: Player }) public players!: MapSchema<Player> = new MapSchema<Player>();
+    @type({ map: Player }) public players: MapSchema<Player> = new MapSchema<Player>();
     @type("boolean") public running!: boolean;
 }
```
... but honestly I'm not so sure if that is enough.

Also, is this possible to generate all the strings within `'  '` instead of `"  "`, and remove trailing whitespaces?

Because after removing the `!`, it compiled normally, but when running `ng lint` (tslint) appears something like:
```log
$ ng lint
Linting "project"...
/path/to/angular/project/src/app/codegen/GameState.ts:1:3
ERROR: 1:3    no-trailing-whitespace  trailing whitespace
ERROR: 4:3    no-trailing-whitespace  trailing whitespace
ERROR: 6:3    no-trailing-whitespace  trailing whitespace
ERROR: 8:66   quotemark               " should be '
ERROR: 9:24   quotemark               " should be '
ERROR: 9:34   semicolon               Missing semicolon
ERROR: 13:11  quotemark               " should be '

/path/to/angular/project/src/app/codegen/Player.ts:1:3
ERROR: 1:3    no-trailing-whitespace  trailing whitespace
ERROR: 4:3    no-trailing-whitespace  trailing whitespace
ERROR: 6:3    no-trailing-whitespace  trailing whitespace
ERROR: 8:66   quotemark               " should be '
ERROR: 12:11  quotemark               " should be '
```

(I put all the generated code within a "codegen" folder)

I'm using `"colyseus.js": "^0.14.1",` within Angular, and `"colyseus": "^0.14.0",` for the game server

Please let me know if there is anything I can help with this.

**Comment by endel** [comment]
Hi @AlexFBP, 

Replacing the double quotes with single quotes in the generated TS code is an easy fix! I think that'd comply with most default linting configs indeed. Here are the imports source code:

https://github.com/colyseus/schema/blob/6213126d4e3cc88469bbb457f5ab078978931b0f/src/codegen/languages/ts.ts#L48-L55

I don't understand why your map of players got generated using `!` though. As you can see in the codegen sources, the `!` is only generated in case there is no initializer/assignment:

https://github.com/colyseus/schema/blob/6213126d4e3cc88469bbb457f5ab078978931b0f/src/codegen/languages/ts.ts#L107-L110

**Comment by AlexFBP** [comment]
Thanks for your help/reply/support @endel . Me neither know too why is generating the `!` :sweat_smile: , I saw that line (107) too :thinking: 

If it helps, what I did was to generate an empty project with `npm init colyseus-app ./my-colyseus-app` and after that, I modified/created the following files:

src/rooms/MyRoom.ts
```typescript
import { Room, Client } from "colyseus";
import { GameState } from "./schema/MyRoomState";

export class GameRoom extends Room {

  onCreate (options: any) {
    console.log("GameRoom created!", options);

    this.setState(new GameState());

    this.onMessage("nextChar", (client, msg) => {
      console.log(client.sessionId, "wants the next character");
      this.state.nextCharacter(client.sessionId);
    });

  }

  onJoin (client: Client, options: any) {
    client.send("ola k ase");
    console.log("Client", client.sessionId, "joined. Options:", options);
    this.state.createPlayer(client.sessionId);
  }

  onLeave (client: Client, consented: boolean) {
    console.log("Client", client.sessionId,"left. Consented:",consented?"yes":"no")
    this.state.removePlayer(client.sessionId);
  }

  onDispose() {
    console.log("Dispose GameRoom");
  }

}
```

src/rooms/schema/MyRoomState.ts
```typescript
import { Schema, type, MapSchema } from "@colyseus/schema";
import { CHARACTERS } from "./Characters";

export class Player extends Schema {

  @type("string")
  character: string = CHARACTERS[Math.floor(Math.random() * CHARACTERS.length)];

}

export class GameState extends Schema {

  @type({ map: Player })
  players = new MapSchema<Player>();

  @type("boolean")
  running: boolean = false;

  createPlayer(sessionId: string) {
    this.players.set(sessionId, new Player());
  }

  removePlayer(sessionId: string) {
    this.players.delete(sessionId);
  }

  nextCharacter (sessionId: string) {
    const old = this.players.get(sessionId).character;
    const next = CHARACTERS.find((v,k) => CHARACTERS[(k-1+CHARACTERS.length) % CHARACTERS.length] === old);
    this.players.get(sessionId).character = next;
  }

}
```

src/rooms/schema/Characters.ts
```typescript
export const CHARACTERS = [
    "zeus",
    "toby",
    "vera"
];
```

Please note that this is mostly a cut down version of the [colyseus-examples](https://github.com/colyseus/colyseus-examples)

**Comment by endel** [comment]
Thanks for the details @AlexFBP, the `!` still bugs me though, I have no idea why it gets generated for you. I've just published a new version of `@colyseus/schema` to use single quotes instead of double quotes for imports, and by using the code you provided, this is the output I get:

```typescript
// 
// THIS FILE HAS BEEN GENERATED AUTOMATICALLY
// DO NOT CHANGE IT MANUALLY UNLESS YOU KNOW WHAT YOU'RE DOING
// 
// GENERATED USING @colyseus/schema 1.0.11
// 

import { Schema, type, ArraySchema, MapSchema, DataChange } from '@colyseus/schema';
import { Player } from './Player'

export class GameState extends Schema {
    @type({ map: Player }) public players: MapSchema<Player> = new MapSchema<Player>();
    @type("boolean") public running!: boolean;
}

```

Can you try upgrading and checking again? Cheers!

**Comment by AlexFBP** [comment]
@endel It generated the same output / worked ! :smiley:

I had to manually add `"@colyseus/schema": "^1.0.11",` to the dependencies section within package.json and `npm i` after that

(`npm update` didn't worked)

Perhaps the package.json of colyseus and colyseus.js could be updated to require ^1.0.10 thus avoiding this issue :sweat_smile: 

Initially I tested 1.0.11 as indicated, but after that, I tested also from 1.0.6 and it got fixed on 1.0.10 (which includes #100, right?)

**Comment by endel** [comment]
I see, that's right! Thanks for sharing!
